Support the SBT extension in BSP import

retronym opened the pull request:
This enables code assist in the `.sbt` and `project/*.scala` files in IntelliJ (pending merge of the [related PR](https://github.com/JetBrains/intellij-scala/pull/593) to intelli-scala).

Other IDEs that support BSP import could also take advantage of this; scala-metals has a ticket to add support: https://github.com/scalameta/metals-feature-requests/issues/143

To disable this feature (e.g. if using an IDE version that does _not_ yet support the [sbt extension](https://build-server-protocol.github.io/docs/extensions/sbt.html) in the BSP protocl, set `ThisBuild / bspSbtEnabled := false`.

Fixes #5953 

Requires: https://github.com/JetBrains/intellij-scala/pull/593

retronym committed:
Support the SBT extension in BSP import

This enables code assist in the .sbt and project/*.scala files
in IntelliJ and any other IDEs that implement this extension.

adpi2 commented on main/src/main/scala/sbt/Keys.scala:
```diff
@@ -399,10 +399,11 @@ object Keys {
   val bspEnabled = SettingKey[Boolean](BasicKeys.bspEnabled)
   val bspTargetIdentifier = settingKey[BuildTargetIdentifier]("Build target identifier of a project and configuration.").withRank(DSetting)
-  val bspWorkspace = settingKey[Map[BuildTargetIdentifier, Scope]]("Mapping of BSP build targets to sbt scopes").withRank(DSetting)
+  val bspWorkspace = settingKey[BspWorkspace]("Mapping of BSP build targets to sbt scopes").withRank(DSetting)
```
In theory that will break the compatibility between sbt versions. In that particular case, I don't think any build or plugin depends on it directly. So it should not be a problem. But maybe we should take the opportunity to make it private?

retronym replied:
It's not strictly necessary to change this key but it sure seemed the cleanest while prototyping. Happy to take direction though about how to proceed to reduce risk of compatibility breakage.

eed3si9n replied:
It's used by at least one plugin - https://github.com/scalacenter/scala-debug-adapter/blob/b3125d8a3873a7218e452b256ed938d14c401c5c/sbt/plugin/src/main/scala/ch/epfl/scala/debugadapter/sbtplugin/DebugAdapterPlugin.scala#L68, but I am guessing we can make exception for that one.

adpi2 replied:
That's right... How did I forget about it? So `bspWorkspace` cannot be private. Also if the signature changes it will be tricky for us, but we can probably work it out anyway.

retronym replied:
I'll rework to avoid a signature change. I can define a new key that aggregates the `Scope` and `LoadedBuildUnit` based targets and leave `val bspWorkspace = settingKey[Map[BuildTargetIdentifier, Scope]]` as a view on the scope part of that data for backwards compatibility.

adpi2 commented on main/src/main/scala/sbt/Keys.scala:
```diff
@@ -399,10 +399,11 @@ object Keys {
   val bspInternalDependencyConfigurations = settingKey[Seq[(ProjectRef, Set[ConfigKey])]]("The project configurations that this configuration depends on, possibly transitivly").withRank(DSetting)
   val bspWorkspaceBuildTargets = taskKey[Seq[BuildTarget]]("List all the BSP build targets").withRank(DTask)
   val bspBuildTarget = taskKey[BuildTarget]("Description of the BSP build targets").withRank(DTask)
+  val bspSbtBuildTarget = taskKey[List[BuildTarget]]("Description of the BSP SBT build targets").withRank(DTask)
```
```suggestion
  val bspSbtBuildTargets = taskKey[List[BuildTarget]]("Description of the BSP SBT build targets").withRank(DTask)
```

retronym replied:
Applied.

adpi2 commented on main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala:
```diff
@@ -133,9 +151,9 @@ object BuildServerProtocol {
       val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
-      val filter = ScopeFilter.in(targets.map(workspace))
+      val workspace = bspWorkspace.value.filter(targets)
+      val filter = ScopeFilter.in(workspace.scopes.values.toList)
```
Would be nice to log a warning if `bspWorkspace.builds` is not empty so that a warn notification is sent to the build client.

Also in scalaMainClasses, scalaTestClasses and test.

retronym replied:
Done.

adpi2 commented on main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala:
```diff
@@ -437,6 +483,50 @@ object BuildServerProtocol {
+      jars = scalaJars.toVector.map(_.toURI.toString)
+    )
+    val workspace = bspWorkspaceSetting.value
+    Def.task {
```
This task does not seem to be needed, in which case the `sbtBuildTargetTask` does not need to be dynamic.

adpi2 commented:
Looks very promising so far. Thank you @retronym for taking care of this!

retronym commented:
I should get back to this later today, thanks for the review comments so far.

The corresponding change to IntelliJ is just about ready to land: https://github.com/JetBrains/intellij-scala/pull/593

retronym committed:
Address review feedback

  - Restore old type of `bspWorkspace` key for backwards compat.
    Instead, introduce `bspFullWorkspace` that includes the
    SBT targets
  - Log a warning if the client requests, e.g. `scalaMainClasses`
    for a SBT target
  - Refactor the code that creates the SBT build targets so it
    doesn't depend on `sbtFullWorkspace`.
  - Add a setting `bspSbtEnabled` to let the user opt-opt of
    SBT target export (e.g. to compensate for a client that does
    not yet support them)

retronym committed:
Add scalacOptions to the SBT targets

retronym commented:
This one is ready for another round of review. I'll try to address attaching SBT source JARs to the project in a followup PR.

eed3si9n commented on main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala:
```diff
@@ -698,4 +824,21 @@ object BuildServerProtocol {
+    }
+    def warnIfBuildsNonEmpty(method: String, log: Logger): Unit = {
+      if (builds.nonEmpty)
+        log.warn(s"$method is a no-op for SBT targets: ${builds.keys.mkString("[", ",", "]")}")
```
```suggestion
        log.warn(s"$method is a no-op for build.sbt targets: ${builds.keys.mkString("[", ",", "]")}")
```

retronym replied:
Applied manually

adpi2 commented on main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala:
```diff
@@ -92,36 +97,54 @@ object BuildServerProtocol {
+          case (id, loadedBuildUnit) =>
+            val base = loadedBuildUnit.localBase
+            val sbtFiles = configurationSources(base)
+            val defDir = projectStandard(base)
```
That's not precise enough because the build client would think that the `project/plugins.sbt` and all the files in `project/project/` directories are source files of the main build target, which is wrong.

Also the `project/src/main/scala` belongs to the sbt build target. And the user is even free to add any other source directory with:
```scala
// project/build.sbt
unmanagedSourceDirectories += baseDirectory.value / "custom"
```
Or even he/she can add some generated sources.

Maybe, a good way to fix this is to pass the all sources from the meta-build to the build context through the `PluginData`.

adpi2 replied:
In other words, we should put more information in here so that the BSP build target is consistent with the meta-build definition:
https://github.com/sbt/sbt/blob/9274a3e365d3cb826ede31598afde412ab6625c1/main/src/main/scala/sbt/internal/Load.scala#L1163-L1174.

retronym replied:
WIP: https://github.com/sbt/sbt/compare/develop...retronym:topic/bsp-import-plugin-data?expand=1

retronym replied:
I've added a commit to this PR that implements your suggestion.

I don't go as far as trying to represent the meta-build (or meta-meta-...-builds!) (`project/*.sbt` / `project/project/*.scala`) as a build target. This _is_ supported by the BSP spec but not yet supported by IntelliJ.

retronym committed:
Record build sources in PluginData for BSP

eed3si9n reviewed (approved):
LGTM, but I yield the detail review to Adrien.

eed3si9n reviewed (dismissed):
LGTM, but I yield the detail review to Adrien.

adpi2 commented on main/src/main/scala/sbt/EvaluateTask.scala:
```diff
@@ -146,12 +146,27 @@ final case class PluginData(
 ) {
   val classpath: Seq[Attributed[File]] = definitionClasspath ++ dependencyClasspath
+
+  // Not a constructor parameter to avoid a binary incompatibility
+  private[this] var _buildSources: BuildSources = BuildSources.Empty
+  private[sbt] def setBuildSourcesData(data: BuildSources): PluginData = {
+    _buildSources = data; this
+  }
+  private[sbt] def buildSourcesData: BuildSources = _buildSources
```
@eed3si9n Is it really necessary to preserve the binary compatibility of `PluginData`? How bad would it be to break it in sbt 1.6?

I am asking this because I also want to add more parameters in it for #6566.

retronym replied:
The breakage is just for the constructor, copy, and apply factory methods, not for consumers of the data. I think it is highly unlikely to break existing plugins.

eed3si9n replied:
>  Is it really necessary to preserve the binary compatibility of `PluginData`? How bad would it be to break it in sbt 1.6?

We maintain binary compatibility for the purpose of maintaining plugin compatibility, so I think `PluginData` would be ok to break, but ultimately we the only way to quantify this would be to release a milestone and test it using Community Build. If none of the existing builds break, then we should be ok.

adpi2 commented on main/src/main/scala/sbt/internal/Load.scala:
```diff
@@ -1164,13 +1164,16 @@ private[sbt] object Load {
         val cp = (Configurations.Runtime / fullClasspath).value
         val opts = (Configurations.Compile / scalacOptions).value
-        PluginData(
+        val managedSrcs = (Configurations.Compile / unmanagedSources).value
+        val unmanagedSrcs = (Configurations.Compile / managedSources).value
```
`managedSrcs` and `unmanagedSrcs` are inverted here

adpi2 commented on main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala:
```diff
@@ -93,27 +99,50 @@ object BuildServerProtocol {
+          case (id, loadedBuildUnit) =>
+            val base = loadedBuildUnit.localBase
+            val sbtFiles = configurationSources(base)
+            val buildSources = loadedBuildUnit.unit.plugins.pluginData.buildSourcesData
```
`buildSources` is always empty because of this `PluginData.copy` here (the private `_buildSources` is not copied):
https://github.com/sbt/sbt/blob/e4231ac03903e174bc9975ee00d34064a1d1f373/main/src/main/scala/sbt/internal/Load.scala#L1252

adpi2 replied:
This would be easier to fix and to maintain if we can break the `PluginData` constructor (@eed3si9n).

adpi2 reviewed (changes requested):
The 2 comments below should be addressed before merging. Otherwise I would suggest to update the test here:
https://github.com/sbt/sbt/blob/e4231ac03903e174bc9975ee00d34064a1d1f373/server-test/src/test/scala/testpkg/BuildServerTest.scala#L24-L35
and to add a new test for the sources of the build.sbt target.

retronym commented:
Update: I'm currently trying to extend the BSP test. Having some difficulty there as one of my new tests for `sources (sbt build)` works in isolation but not after the exising test for normal sources. Haven't figured out the crosstalk yet.

retronym committed:
Refine and test SBT BSP target

  - Pass source dirs and current list of files
  - Align display name and URI

retronym committed:
Exclude create/copy methods of PluginData from MiMa's consideration

retronym committed:
Fix compilation error in test

adpi2 committed:
[BSP] Ignore sbt target in resources request

adpi2 reviewed (approved):
I just pushed a small fix for the new `buildTarget/resources` request. Otherwise it looks good to me as it is.
I tried in Metals and most of the things (goto definitions and code completions) are already working out-of-the-box. That's cool!

retronym commented:
@eed3si9n Are you happy with this one now, in particular deeming `PluginData` creation/copying as sbt-internal and excluded from MiMa?

retronym commented:
There was a scalacheck test failure in one of the test runs, not related to this PR, but maybe worth looking into:

```
[info] - sbt.ParseKey.An unspecified project axis resolves to the current project: Falsified after 4341 passed tests
[info] > StructureKeyMask(Env:
[info]     Tasks:
[info]     aa (delegates: )
[info] Build file:///aaaaaaa/#a :
[info]     Project a
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info]     Project b
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info]     Project aa
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info] current: ProjectRef(file:///aaaaaaa/#a,a)
[info] Settings:
[info] 	Global / <key>
[info] 	   aa
[info] All keys:
[info] 	aa,ScopedKey(Scope(Select(BuildRef(file:///aaaaaaa/#a)), Select(ConfigKey(A)), Select(aa), Zero),aa),ScopeMask(false,false,true,false))
[info] > === Not Equal ===
[info] > --- lhs ---
[info] > Select(ConfigKey(A))
[info] > --- rhs ---
[info] > Zero
[info] > === Not Equal ===
[info] > --- lhs ---
[info] > Select(ProjectRef(file:///aaaaaaa/#a,aa))
[info] > --- rhs ---
[info] > Select(ProjectRef(file:///aaaaaaa/#a,a))
[info] > parsed subproject: Select(ProjectRef(file:///aaaaaaa/#a,aa))
[info] > current subproject: ProjectRef(file:///aaaaaaa/#a,a)
[info] > Key: {file:///aaaaaaa/#a} / A / aa / aa
[info] > Mask: ScopeMask(false,false,true,false)
[info] > Key string: 'aa / aa'
[info] > Parsed: Right(ProjectRef(uri("file:///aaaaaaa/#a"), "aa") / aa)
[info] > Structure: Env:
[info]     Tasks:
[info]     aa (delegates: )
[info] Build file:///aaaaaaa/#a :
[info]     Project a
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info]     Project b
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info]     Project aa
[info]       Delegates:
[info]         
[info]       Configurations:
[info]         A
[info] current: ProjectRef(file:///aaaaaaa/#a,a)
[info] Settings:
[info] 	Global / <key>
[info] 	   aa
[info] All keys:
[info] 	aa
```

eed3si9n commented:
@retronym I'm guessing it's a known ambiguity when we have subproject named `aa` and a task key `aa`, `aa / aa` can be interpreted as "`aa` task in `aa` project", not "`aa` task scoped to `aa` task in current subproject", and thus "An unspecified project axis resolves to the current project" gets falsified. I probably should try to filter that out of the test.

I'm fine with `PluginData`. I guess the question is if we should call this a 1.6.0 feature or 1.5.x. We can publish 1.6.0-M1, run Community Build, and then back port it to 1.5.x if it passes?

retronym commented:
I did a GitHub code search for "plugindata language:scala", filtered out files with `package sbt`, and didn't spot any usages of the changed methods.

```
$ gh api --paginate -X GET search/code -f q='language:scala plugindata' -q ".items[] | .url" > /tmp/urls.txt
$ for url in $(cat /tmp/urls.txt); do echo $url; code=$(gh api $url  -q ".content" | base64 --decode); echo $code | grep 'package sbt' || echo $code; done | tee /tmp/code.txt
```

Not a perfect method, but I think it is likely safe enough for 1.5.x. @adpi2 mentioned that he also has a followup change to add extra data to `PluginData`.

eed3si9n commented:
ok. We can merge it now, and see.
